ProjFS.Mac/PrjFSLib/prjfs-log/prjfs-log.cpp

```diff
@@ -10,41 +10,115 @@
 static const char* KextLogLevelAsString(KextLog_Level level);
 static uint64_t NanosecondsFromAbsoluteTime(uint64_t machAbsoluteTime);
 static dispatch_source_t StartKextProfilingDataPolling(io_connect_t connection);
+static void ProcessLogMessagesOnConnection(io_connect_t connection, io_service_t prjfsService);
 
 static mach_timebase_info_data_t s_machTimebase;
-
+static uint64_t s_machStartTime;
+static IONotificationPortRef s_notificationPort;
 
 int main(int argc, const char * argv[])
 {
     mach_timebase_info(&s_machTimebase);
-    const uint64_t machStartTime = mach_absolute_time();
+    s_machStartTime = mach_absolute_time();
+
+
+    s_notificationPort = IONotificationPortCreate(kIOMasterPortDefault);
+    IONotificationPortSetDispatchQueue(s_notificationPort, dispatch_get_main_queue());
 
-    io_connect_t connection = PrjFSService_ConnectToDriver(UserClientType_Log);
-    if (connection == IO_OBJECT_NULL)
+    PrjFSService_WatchContext* watchContext = PrjFSService_WatchForServiceAndConnect(
+        s_notificationPort, UserClientType_Log,
+        [](io_service_t service, io_connect_t connection, PrjFSService_WatchContext* context)
+        {
+            if (connection != IO_OBJECT_NULL)
+            {
+                ProcessLogMessagesOnConnection(connection, service);
+            }
+        });
+    if (nullptr == watchContext)
     {
-        std::cerr << "Failed to connect to kernel service.\n";
+        std::cerr << "Failed to register for IOService notifications.\n";
         return 1;
     }
     
-    DataQueueResources dataQueue = {};
-    if (!PrjFSService_DataQueueInit(&dataQueue, connection, LogPortType_MessageQueue, LogMemoryType_MessageQueue, dispatch_get_main_queue()))
+
+    CFRunLoopRun();
```

Comment:
* Calling `CFRunLoopStop()` on it from anywhere will stop it. Currently, nothing calls that, as we don't have any events that would cause our tool to exit. If we turn this into a daemon, we might want to handle some signals for exiting cleanly.
 * As this is the "main" runloop, events for all the `dispatch_source` objects associated with the "main" dispatch queue (search for `dispatch_get_main_queue()`) are being handled here. CFRunLoops and Grand Central Dispatch (libdispatch) aren't quite the same thing but they overlap on their "main" runloop and dispatch queue. Instead of calling `CFRunLoopRun()` we could also use `dispatch_main()` but that call will never return, so I prefer the former.

Reply:
> Will `PrjFSService_StopWatching` (the call after `CFRunLoopRun`) get hit? (e.g. if the logger gets a `SIGTERM`)

The default SIGTERM handler doesn't call `CFRunLoopStop()`, no. Since all resources we currently use are of the kind that clean themselves up when the process dies, this isn't a problem, but if that changes, we'll want that `PrjFSService_StopWatching()` there, and having it there now is a good reminder.